### PR TITLE
Feature/165 keycloack

### PR DESCRIPTION
**Pull request by filip-kopecky**
Added keycloak option to the UI. When passed` REACT_APP_AUTHENTICATION=keycloak` in .env the UI is using that auth provider

### CODE REVIEW

File: src/reducer/TermItReducers.ts [hunk 1]
```diff
@@ -95,6 +95,8 @@ function messages(state: Message[] = [], action: MessageAction): Message[] {
     case ActionType.LOGOUT:
     case ActionType.LOGIN: // Intentional fall-through
       return [];
+    case ActionType.LOGIN_KEYCLOAK: // Intentional fall-through
```
Comment (ledsoft) [inline src/reducer/TermItReducers.ts:97]: There is no fall-through here. Either remove the comment, or modify the code so that the comment corresponds to reality.

Comment (filip-kopecky) [reply src/reducer/TermItReducers.ts:97]: Sorry about that, I was copying code without much thinking I suppose

File: src/action/AsyncUserActions.ts [hunk 2]
```diff
@@ -42,14 +44,18 @@ export function loadUser() {
...
     return Ajax.get(`${Constants.API_PREFIX}${USERS_ENDPOINT}/current`)
-      .then((data: object) =>
-        JsonLdUtils.compactAndResolveReferences<UserData>(data, USER_CONTEXT)
-      )
+      .then((data: object) => {
+        return JsonLdUtils.compactAndResolveReferences<UserData>(
+          data,
+          USER_CONTEXT
+        );
+      })
```
Comment (ledsoft) [inline src/action/AsyncUserActions.ts]: Unnecessary return results in unnecessary change in formatting.

File: src/component/login/LoginKeycloak.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      </Card>
+    );
+  }
+};
+
+export default connect(undefined, (dispatch: ThunkDispatch) => {
+  return {
+    login: () => dispatch(loginKeycloak),
+  };
+})(injectIntl(withI18n(Login)));
```
Comment (ledsoft) [inline src/component/login/LoginKeycloak.tsx]: Please, rewrite to `useDispatch` and `useI18n` hooks so that we can get rid of the `connect` HOC.

File: src/component/misc/AuthUnavailable.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+const AuthUnavailable: React.FC<HasI18n> = ({ i18n }) => {
+  return (
+    <Card>
+      <CardHeader tag="h2">{Constants.APP_NAME}</CardHeader>
+      <CardBody>{i18n("auth.unavailable-message")}</CardBody>
+    </Card>
+  );
+};
+
+export default injectIntl(withI18n(AuthUnavailable));
```
Comment (ledsoft) [inline src/component/misc/AuthUnavailable.tsx]: Please, rewrite to the `useI18n` hook and remove the HOCs.

File: src/reducer/TermItReducers.ts [hunk 5]
```diff
@@ -95,6 +95,8 @@ function messages(state: Message[] = [], action: MessageAction): Message[] {
     case ActionType.LOGOUT:
     case ActionType.LOGIN: // Intentional fall-through
       return [];
+    case ActionType.LOGIN_KEYCLOAK:
+      return [];
```
Comment (ledsoft) [inline src/reducer/TermItReducers.ts]: We can probably let it fall-through to the same action, as the result is the same anyway - returning an empty array. So something like
```javascript
case ActionType.LOGOUT:
case ActionType.LOGIN: // Intentional fall-through
case ActionType.LOGIN_KEYCLOAK: // Intentional fall-through
  return [];
```

File: src/component/MainView.tsx [hunk 6]
```diff
@@ -64,6 +65,7 @@ export class MainView extends React.Component<MainViewProps, MainViewState> {
 
   public componentDidMount(): void {
     if (this.props.user === EMPTY_USER) {
+      Routing.saveOriginalTarget();
```
Comment (ledsoft) [inline src/component/MainView.tsx:70]: Could just confirm that this is because Keycloak-based backend returns 302 forward to Keycloak login instead of 401? Because for 401 responses this is handled by Ajax.

**Review by ledsoft — CHANGES_REQUESTED** [review]
See inline comments (but otherwise I tested it and it works).

**Review by ledsoft — APPROVED** [review]
LGTM
